irradiance_qa/irradiance_qa.py

```diff
@@ -0,0 +1,670 @@
+
+    # read the data created by process function
+    # do this first so that we don't attempt to modify data in Arbiter unless
+    # we know this is good. The cost of safety is the time and memory used to
+    # read approximately 3*10*20MB = 600MB of csv data.
+    logger.info('reading site data')
+    data_to_post = defaultdict(dict)
+    for site_name in SITES:
+        p = OUTPUT_PATH / f'{site_name}' / 'validated_data'
+        for variable in ('ghi', 'dni', 'dhi'):
+            f = p / f'{site_name}_validated_or_sfa_user_flagged_{variable}.csv'
+            logger.debug('reading %s', f)
+            data = pd.read_csv(f, index_col=0, parse_dates=True)
+            if not (data.columns == pd.Index(['value', 'quality_flag'])).all():
+                raise ValueError(f'wrong columns in {f}')
+            if not isinstance(data.index, pd.DatetimeIndex):
+                raise ValueError(f'wrong index in {f}')
+            data_to_post[site_name][variable] = data
+
+    token = cli_access_token(user, password)
+    session = APISession(token, base_url=base_url)
+    sites = session.list_sites()
+    ta23_sites = tuple(filter(lambda x: x.name in SITES, sites))
+    obs = session.list_observations()
+    def _is_reference_obs(o):
+        return (
+            o.site in ta23_sites and
+            o.variable in ['ghi', 'dni', 'dhi'] and
+            o.site.provider == 'Reference'
+        )
+    ta23_obs = tuple(filter(_is_reference_obs, obs))
+    if official:
+        # use the real obs
+        obs_for_post = ta23_obs
+    else:
+        # Create new obs patterned on real obs.
+        # Same as ta23_obs but have new uuids and provider.
+        # (uuid and provider is set by SFA API)
+        obs_for_post = [
+            session.create_observation(o) for o in ta23_obs
+        ]
+    for o in obs_for_post:
+        site_name = o.site.name
+        variable = o.variable.lower()
+        _data_to_post = data_to_post[site_name][variable]
+        # Split into chunks to stay under API upload limit.
+        grouped_data = _data_to_post.groupby(lambda x: (x.year, x.month))
+        for yr_mo, values in grouped_data:
+            logger.debug(
+                'posting data for %s %s %s', yr_mo, site_name, variable
+            )
+            session.post_observation_values(o.observation_id, values)
+            # sometimes API fails to validate upload, so wait and retry if
+            # needed
+            validated = wait_for_validation(session, o, values)
+            if not validated:
+                logger.info('validation appears hung. reposting data')
+                session.post_observation_values(o.observation_id, values)
+                validated = wait_for_validation(session, o, values)
+            if not validated:
+                logger.info('validation appears hung. reposting data')
+                session.post_observation_values(o.observation_id, values)
+                validated = wait_for_validation(session, o, values)
```

Comment:
Do you want to do this twice? Seems like wait_for_validation does some amount of retrying?